### ISSUE TITLE
Filtered counts from remove-background higher than raw input counts

### ISSUE THREAD

**Issue by colin-umn** [broadinstitute/CellBender#40]
I am seeing counts filtered by remove-background actually get higher relative to the raw counts that I feed into remove-background.  Is this to be expected given the algorithms involved?  Is the idea that it is relative counts that matter and so increasing counts is okay and maybe even expected.

The overall gene counts per cell tend to go up by factors around 1.5 to 2.5.  The unique gene counts per cell mostly decrease, but even some of these increase.

**Comment by johnchamberlin** [comment]
I have observed the same. In my case, the increase in total counts was only 0.7%. It would be helpful to see some metrics in the output log describing how many counts were removed, etc.

**Comment by colin-umn** [comment]
Here is the log.   I have little experience running samples through CellBender.  I suppose the barcode counts in the log might indicate a quality problem.

cellbender:remove-background: Command:
cellbender remove-background --input LB_PBMC3/outs/raw_feature_bc_matrix.h5 --output xp_300c.h5 --epochs 300 --cuda --z-dim 200 --z-layers 1000
cellbender:remove-background: 2019-11-03 09:05:46
cellbender:remove-background: Running remove-background
cellbender:remove-background: Loading data from file LB_PBMC3/outs/raw_feature_bc_matrix.h5
cellbender:remove-background: CellRanger v3 format
cellbender:remove-background: Trimming dataset for inference.
cellbender:remove-background: Prior on counts in empty droplets is 1210
cellbender:remove-background: Prior on counts for cells is 6466
cellbender:remove-background: Excluding barcodes with counts below 968
cellbender:remove-background: Using 236 probable cell barcodes, plus an additional 4539 barcodes, and 0 empty droplets.
cellbender:remove-background: Running inference...
cellbender:remove-background: [epoch 000]  average training loss: 2670.4840
...
cellbender:remove-background: [epoch 299]  average training loss: 2398.1941
cellbender:remove-background: Inference procedure complete.
cellbender:remove-background: Preparing to write outputs to file...
cellbender:remove-background: Encoding data according to model.
cellbender:remove-background: Getting ambient-background-subtracted UMI count matrix.
cellbender:remove-background: Succeeded in writing output to file xp_300c.h5
cellbender:remove-background: Succeeded in writing output to file xp_300c_filtered.h5
cellbender:remove-background: Saved cell barcodes in xp_300c_cell_barcodes.csv
cellbender:remove-background: Saved summary plots as xp_300c.pdf
cellbender:remove-background: Completed remove-background.
cellbender:remove-background: 2019-11-03 09:18:02

**Comment by johnchamberlin** [comment]
To clarify, I meant that as a suggestion to the developers (i am not affiliated). As you show, the log produced as-is does not quantify total UMIs. I have only just begun testing CellBender myself.

What was the total UMIs in your filtered output was input? Perhaps the program is trying to impute the "true" expression profile, which would explain why my data did not appreciably change in total UMIs. Hopefully the developers can chime in soon.

**Comment by sjfleming** [comment]
@colin-umn @johnchamberlin You have identified an important caveat to our current algorithm!

The algorithm as it stands does a bit of imputation.  That is to say, it attempts to reconstruct the "true" denoised counts for each cell.  It does this using an auto-encoder setup.  But in the current version, there is no guarantee in the model that the output counts will be strictly less than the input.  As you say, the relative counts should be preserved, and the overall counts should be close to the input.  (If you see a big deviation in counts of a particular gene, this could mean that the auto-encoder is over-parameterized.  Try reducing z-dim and z-layers.)

We are currently developing a "version 2", with a slightly different model, where this requirement is explicitly built in.  There will be no imputation, and the output counts will be strictly less than the input.  You will be able to see exactly which counts were removed.  This is still under development, but we are hoping to release this version in a few months.

**Comment by sjfleming** [comment]
Update: this is currently implemented in the `sf_removebkg_v2.1` branch.  Feel free to try it out.  It's not set in stone yet, but we plan to get it 100% stable in the next month or so and make this the official 0.2 version of the tool.
